### PR TITLE
HAL refactoring: RCC and GPIO modules

### LINKED ISSUE #1171: Factor out GPIO interface in HAL
**What has to be done**
The GPIO interface in the HAL appears to be something that can be delineated and factored out into a class of its own.
See the HAL section of this [document](https://docs.google.com/document/d/1hztIBu9DR52r0LQuIpt6DZByBLfXfJy21BGqk2F34g8/edit) about our architecture.

**How do you know it has to be done**
Much of the HAL is still untested and untestable. Some of this is unavoidable because of how close to the hardware it is, but it should be possible to encapsulate some higher level functions and then mock the lower layers.

Furthermore, the HAL is very monolithic and it is impossible to initialize just one subset of it without mobilizing the entire orchestra. It is worth carving it up just make it easier to read and think about, but if things can be initialized piecemeal, it would be much easier to integration-test subsystems (with hardware in the loop or even just eletronic harnesses) and know with much more clarity which subdivisions of the HAL have been tested.

**Starting points**
As the i2c interface implementation has proven, it is possible to make some HAL susbsystems mockable and testable. We should follow this pattern and improve on it.

**Don't forget**
- [ ] Put the issue in a milestone
- [ ] Assign it to one of the projects
- [ ] Add appropriate labels
- [ ] Mention dependent/blocked issues if any

### PR DESCRIPTION
**Pull request by martukas**
# Description

This factors out the following 2 components out of the STM32 HAL:
* Reset and clock control
* GPIO

and updates other HAL modules accordingly. Although the code is not unit-testable, it can technically compile and link under the `native` environment, which will mean it can be counted under test coverage. This at least gives us honest awareness of the situation.

Also, this moves in the direction of allowing various modules to be implemented with more finely defined dependencies, rather than just have them bring in the entirety of the STM32 HAL in for the ride. This is now at least true for GPIO, which only depends on `clock.h`. To get the other modules similarly cleaned up, we will need to factor out a few additional parts of the HAL, such as timer and ADC registers. I am submitting this PR now so as to not have it blow up, but will continue this work after some feedback.

Closes #1171

# General checklist:

- [x] I have performed a self-review, including - looked through the `Files changed` tab, browsed repository in my branch
- [x] I have made corresponding changes to the documentation - to reflect changes in code, electrical or mechanical design
- [x] All new documentation or graphics follow the [documentation style guide](https://github.com/RespiraWorks/Ventilator/wiki/Documentation-style-guide)
- [x] All new content is linked, easily discoverable, does not require too many clicks
- [x] I have reviewed any other relevant parts of our [contributor wiki](https://github.com/RespiraWorks/Ventilator/wiki) to ensure that this PR conforms to the standards
- [x] I have given the PR a descriptive name (prefixed with **WIP** if it is not yet ready for review)
- [x] I have tagged relevant reviewers

## Code checklist:

- [x] All new code follows our [code style](https://github.com/RespiraWorks/Ventilator/wiki/Code-style)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have described tests I used to verify my changes, instructions for reproducing them and any relevant configuration details.
- [x] I ran the unit test suite and verified that all tests pass - new and old, locally and on CI
- [x] I performed a clean build and verified that there were no warnings
- [x] I ran our static analysis tools and verified there were no warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] All source files have license headers

### CODE REVIEW

File: software/controller/lib/hal/gpio.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+#include <cstdint>
+
+// General Purpose I/O [RM] 8.4
+
+namespace GPIO {
+
+/// \TODO: Move these mappings elsewhere
+enum class Port : uint32_t {
+  PortA = 0x48000000,
```
Comment (asmodai27) [inline software/controller/lib/hal/gpio.h:33]: IMO, `GPIO::Port::A` rolls of better than `GPIO::Port::PortA`, though I agree it's better to have the enum names themselves be explicit (especially when `using` exists).

**Review by asmodai27 — APPROVED** [review]
A nice start to refactoring the HAL, with more explicit stuff and inter-dependencies starting to appear more clearly.
Tested this on my unit yielded no visible regression, with simple qualitative tests:

- trace readings from all sensors with reasonable dynamic
- cycle the pinch valves and blower power, check the resulting flow readings
- standard run in pressure control mode
- EEPROM read and write

File: software/controller/lib/hal/adc.cpp [hunk 2]
```diff
@@ -135,8 +138,8 @@ static constexpr int AdcConversionTime = [] {
 }();
 
 // Calculate how long our history buffer needs to be based on the above.
-static constexpr int AdcSampleHistory = static_cast<int>(
-    SampleHistoryTimeSec * CPU_FREQ / AdcConversionTime / OversampleCount / AdcChannels);
+static constexpr int AdcSampleHistory = static_cast<uint32_t>(
```
[the remark this answers is not in the document]
Comment (martukas) [reply software/controller/lib/hal/adc.cpp:141]: Good catch

File: software/controller/lib/hal/clocks.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+limitations under the License.
+*/
+
+#pragma once
+
+#include <cstdint>
+
+// Please read manual carefully, as not all peripherals ara available on all processors
+// [RM] 6.4.9-26 (pg 208-234)
+enum class PeripheralID {
```
[the remark this answers is not in the document]
Comment (martukas) [reply software/controller/lib/hal/clocks.h:22]: Yes, these possibly belong elsewhere, but at this point I do not have a clear enough picture. Maybe we can do this at a later refactoring step?

File: software/controller/lib/hal/gpio.h (hunk 1 above)
Comment (martukas) [reply software/controller/lib/hal/gpio.h:33]: Well, `using` should be used very sparingly anyhow, so actually I think you have a point there. In fact, if it was less wordy, there would be less of a reason for `using`.

File: software/controller/lib/hal/clocks.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+inline RccReg *const RccBase = reinterpret_cast<RccReg *>(0x40021000);
+
+struct PeripheralMapping {
+  PeripheralBlock offset;
+  uint8_t bit;
+};
+
+PeripheralMapping get_peripheral_mapping(const PeripheralID id) {
+  switch (id) {
+    // [RM] 6.4.9, 6.4.15, 6.4.21
```
[the remark this answers is not in the document]
Comment (martukas) [reply software/controller/lib/hal/clocks.cpp:68]: Yes, one-liners would be nicer, but I'll move ahead with functional refactoring and we can try to make things read easier once it takes shape

**Pull request merged**